### PR TITLE
batch compat old server

### PR DESCRIPTION
**Pull request by stuBirdFly**
## Summary

batch compat old server

## Solution Description

Compatible with old server versions, determine the server version to decide which batch logic to use.

### CODE REVIEW

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 1]
```diff
@@ -540,38 +543,175 @@ public boolean[] exists(List<Get> gets) throws IOException {
...
+                throw new FeatureNotSupportedException(
+                        "not supported other type in batch yet,only support put and delete");
+            }
+        }
+        List<Integer> resultMapSingleOp = new LinkedList<>();
+        List<Row> puts = new LinkedList<>();
+        for (int i = 0; i < actions.size(); i++) {
+            BatchOperationResult tmpResults;
+            BatchOperation batch = null;
+            if (actions.get(i) instanceof Put && ((Put)actions.get(i)).getFamilyCellMap().size() == 1) {
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:559]: this if statement is too long, would you use a local variable to replace (Put)actions.get(i)).getFamilyCellMap().size()?

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 2]
```diff
@@ -540,38 +543,175 @@ public boolean[] exists(List<Get> gets) throws IOException {
...
+                puts.clear();
+                Put put = (Put) actions.get(i++);
+                String family = Bytes.toString(put.getFamilyCellMap().firstKey());
+                // aggregate put
+                puts.add(put);
+                while (i < actions.size()) {
+                    if (actions.get(i) instanceof Put && ((Put)actions.get(i)).getFamilyCellMap().size() == 1 && Objects.equals(
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:566]: same as above

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 3]
```diff
@@ -540,38 +543,175 @@ public boolean[] exists(List<Get> gets) throws IOException {
...
+        List<Integer> resultMapSingleOp = new LinkedList<>();
+        List<Row> puts = new LinkedList<>();
+        for (int i = 0; i < actions.size(); i++) {
+            BatchOperationResult tmpResults;
+            BatchOperation batch = null;
+            if (actions.get(i) instanceof Put && ((Put)actions.get(i)).getFamilyCellMap().size() == 1) {
+                puts.clear();
+                Put put = (Put) actions.get(i++);
+                String family = Bytes.toString(put.getFamilyCellMap().firstKey());
+                // aggregate put
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:563]: this function is too long, can we wrapper this put aggregation process into a internal function

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 4]
```diff
@@ -1691,68 +1835,68 @@ private KeyValue modifyQualifier(Cell original, byte[] newQualifier) {
...
+                if (put.isEmpty()) {
+                    throw new IllegalArgumentException("No columns to insert for #"
+                            + (posInList + 1) + " item");
+                }
+                for (Map.Entry<byte[], List<Cell>> entry : put.getFamilyCellMap().entrySet()) {
+                    byte[] family = entry.getKey();
+                    List<Cell> keyValueList = entry.getValue();
+                    for (Cell kv : keyValueList) {
+                        singleOpResultNum++;
+                        if(isTableGroup){
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:1862]: can we handle tablegroup requset in buildMutation like Delete ?

Comment (stuBirdFly) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: fixed

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 3 above)
Comment (stuBirdFly) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: fixed

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 2 above)
Comment (stuBirdFly) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: Must check whether it is a put before calling getFamilyMapCell. If using a local variable, an additional if statement needs to be nested inside the existing if

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 1 above)
Comment (stuBirdFly) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: Must check whether it is a put before calling getFamilyMapCell. If using a local variable, an additional if statement needs to be nested inside the existing if

**Pull request merged**